src/cpp/session/modules/rmarkdown/NotebookExec.cpp

```diff
@@ -0,0 +1,232 @@
+
+#include <core/text/CsvParser.hpp>
+#include <core/FileSerializer.hpp>
+
+#include <session/SessionModuleContext.hpp>
+#include <session/SessionUserSettings.hpp>
+#include <session/SessionSourceDatabase.hpp>
+
+#include <iostream>
+
+using namespace rstudio::core;
+
+namespace rstudio {
+namespace session {
+namespace modules {
+namespace rmarkdown {
+namespace notebook {
+
+namespace {
+
+bool moveLibFile(const FilePath& from, const FilePath& to, 
+      const FilePath& path, std::vector<FilePath> *pPaths)
+{
+   std::string relativePath = path.relativePath(from);
+   FilePath target = to.complete(relativePath);
+
+   pPaths->push_back(path);
+
+   Error error = path.isDirectory() ?
+                     target.ensureDirectory() :
+                     path.move(target);
+   if (error)
+      LOG_ERROR(error);
+   return true;
+}
+
+} // anonymous namespace
+
+ChunkExecContext::ChunkExecContext(const std::string& docId, 
+      const std::string& chunkId):
+   docId_(docId), 
+   chunkId_(chunkId),
+   connected_(false)
+{
+}
+
+ChunkExecContext::~ChunkExecContext()
+{
+   if (connected_)
+      disconnect();
+}
+
+std::string ChunkExecContext::chunkId()
+{
+   return chunkId_;
+}
+
+std::string ChunkExecContext::docId()
+{
+   return docId_;
+}
+
+bool ChunkExecContext::connected()
+{
+   return connected_;
+}
+
+void ChunkExecContext::connect()
+{
+   FilePath outputPath = chunkOutputPath(docId_, chunkId_);
+   Error error = outputPath.ensureDirectory();
+   if (error)
+   {
+      // if we don't have a place to put the output, don't register any handlers
+      // (will end in tears)
+      LOG_ERROR(error);
+      return;
+   }
+
+   // begin capturing plots 
+   connections_.push_back(events().onPlotOutput.connect(
+         boost::bind(&ChunkExecContext::onFileOutput, this, _1, 
+                     kChunkOutputPlot)));
+
+   error = beginPlotCapture(outputPath);
+   if (error)
+      LOG_ERROR(error);
+
+   // begin capturing HTML input
+   connections_.push_back(events().onHtmlOutput.connect(
+         boost::bind(&ChunkExecContext::onFileOutput, this, _1, 
+                     kChunkOutputHtml)));
+
+   error = beginWidgetCapture(outputPath, 
+         outputPath.parent().complete(kChunkLibDir));
+   if (error)
+      LOG_ERROR(error);
+
+   // begin capturing console text
+   connections_.push_back(module_context::events().onConsolePrompt.connect(
+         boost::bind(&ChunkExecContext::onConsolePrompt, this, _1)));
+   connections_.push_back(module_context::events().onConsoleOutput.connect(
+         boost::bind(&ChunkExecContext::onConsoleOutput, this, _1, _2)));
+   connections_.push_back(module_context::events().onConsoleInput.connect(
+         boost::bind(&ChunkExecContext::onConsoleInput, this, _1)));
+
+   connected_ = true;
+}
+
+void ChunkExecContext::onConsolePrompt(const std::string& )
+{
+   if (connected_)
+      disconnect();
+}
+
+void ChunkExecContext::onFileOutput(const FilePath& file, int outputType)
```

Comment:
Not at present. It isn't necessary for the cache (which is all we're populating here) but will be necessary for the .Rnb itself when we start serializing that.